### PR TITLE
Prevent lookup of partition ID on LiteMembers with Near Cache

### PR DESCRIPTION
**Pull request by Donnerbart**
With the key store-by-reference feature for the Near Cache the
`NearCachedMapProxyImpl.isOwn()` method introduced an unwanted
serialization of the key to check if a key belongs to this member.
Lite Members can skip this check, since they never have local keys.

This PR is needed to show the effect of Near Cache keys stored by-reference.
If we don't merge that feature this PR will show not much effect.

Depends on https://github.com/hazelcast/hazelcast/pull/10555

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java [hunk 1]
```diff
@@ -73,6 +75,8 @@
     public NearCachedMapProxyImpl(String name, MapService mapService, NodeEngine nodeEngine, MapConfig mapConfig) {
         super(name, mapService, nodeEngine, mapConfig);
 
+        localMember = getNodeEngine().getLocalMember();
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java]: You should not cache localMember. When lite member is promoted, `localMember` field of ClusterService is updated.

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java [hunk 2]
```diff
@@ -508,7 +512,7 @@ private long tryReserveForUpdate(Data key) {
     }
 
     private boolean cachingAllowedFor(Data key) {
-        return cacheLocalEntries || !isOwn(key);
+        return cacheLocalEntries || localMember.isLiteMember() || !isOwn(key);
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java]: Should be `getNodeEngine().getLocalMember().isLiteMember()`.

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java (hunk 1 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java]: It seems `NodeEngine.getLocalMember()` javadoc is not correct anymore:

```
/**
     * Returns the local member.
     * <p/>
     * The returned value will never change and will never be null.
     *
     * @return the local member.
     */
    Member getLocalMember();
```

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java (hunk 2 above)
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java]: The `getNodeEngine()` seems to have some overhead of a lifecycle check. Also `getLocalMember()` is redirected several times. The `ClusterService` seems to be the source of truth, so I picked `clusterService.getLocalMember().isLiteMember()` and cache the `ClusterService` as field. I hope this is the most performant and correct solution :)

**Comment by Donnerbart** [comment]
With the Near Cache key store-by-reference the keys can be in Object or Data format. But the `isOwn()` check needs a serialized key for the partition lookup. So there will be an on-demand serialization in that case. This PR prevents this at least for Lite Members.
screenshot from 2017-05-17 11-39-00

**Comment by Donnerbart** [comment]
I've added this small fix into https://github.com/hazelcast/hazelcast/pull/10555

**Pull request closed without merging**